README.md
[![ko-fi](https://www.ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/D1D02SYZA)

# travel-pace
A simple module that calculates the travel time for D&amp;D5e players. it's just a module I'm using to learn how to make modules for foundry.

- Use the information described in pg.84 of the OGL as a calculation rule.
- (tips) Use the lowest speed of the group in feet as a reference.
- (tips) It is implicit that you consider any terrain that does not consist of a road or trail as difficult terrain.

# Manifest
`https://raw.githubusercontent.com/rinnocenti/travel-pace/master/module.json`

## How to Use

### Settings
<img src="https://github.com/rinnocenti/travel-pace/blob/master/img/readme01.jpg" width="70%" height="70%">

#### Forced March on Chat
By activating this setting, a forced march rules message will be sent to the chat to alert players of the forced march rules. Disabling this message is not displayed in the chat.

#### Convert to Metric System
if you have a module that changes the measurement values for the metric system, you can also convert the module to the system, enabling this field

### On Map
<img src="https://github.com/rinnocenti/travel-pace/blob/master/img/readme02.jpg" width="70%" height="70%">
In the basic control bar there will be a button called Travel Pace, when clicking a dialog screen will appear.

#### Lowest group speed feets
Enter the speed of the group.
Ideally, it should be the lowest speed among everyone in the group.

#### Miles on the road/trails
Enter here the distance traveled by the players on the roads or trails.
Basically this field considers any movement on NOT difficult terrain

#### Miles on the road/trails
Enter here the distance traveled by the players off the roads or trails.
Basically this field considers any movement on the difficult terrain

#### Speed Ratio
sometimes on certain types of land speed is further reduced, usually the players do three miles per hour. however, this average can be changed according to the ground rules.
* Normal (SRD) -> 3 mi/h
* 1/2 Normal -> 1.5 mi/h
* 1/4 Normal -> 0.75 mi/h
* 1/6 Normal -> 0.5 mi/h - ( rules of Icewind Dale: Rime of the Frostmaiden)

#### Estimated Hours
It will display a time estimate for the displayed configuration.It will display a time estimate for the displayed configuration. Presenting the time in the 3 gears, Normal, fast and slow, together with the rules of each one.

#### Forced March
A summary of the forced march rules, if you selected in the settings to display forced march, this text will be sent together with the proposed march for the chat.

#### Buttons
decided which type of gear will be used the player / master presses one of the buttons and a summary of the choices will be sent to the chat.

### On Chat
<img src="https://github.com/rinnocenti/travel-pace/blob/master/img/readme03.jpg" width="50%" height="50%">
A message will be sent to the chat with a description of what was chosen. containing some information:

* Who measure.
* With rate used.
* Many miles in Dificult and Normal terrain
* Estimatade time
* Days for compleat the journey (by rule players can only travel for 8h a day, without the use of forced march.)
* Nonus and Penalties
* (Optional) Rules of Forced March.

## Support
If you like this module and would like to help or found a bug or request new features call me on discord @Innocenti#1455 or create a issue here.

## License
This Foundry VTT module, writen by Innocenti, is licensed under a Creative Commons Attribution 4.0 International License.



# Feedback
For bugs/feedback, create an issue on GitHub, or contact me on Discord at Innocenti#1455


CHANGELOG.md
Updates

v.2.1.3: Update for V11

v.1.1.3: Update save "0" values

v.1.1.2: Update for V10 (work in v8,v9 and v10)

v.1.1.1: EN.lang update

v.1.1.0: Final update

v.0.5.0: Update for 0.8.6
- Persistent values.
- BREAKING: REINSTALL IS NECESSARY
- WORK IN 0.7.9 and 0.8.6

v 0.3.0: added possibility of different speed ratios in accordance with the campaign or local.
- Normal pace (3 unities per hour).
- 1/2 (1.5 unities per hour)
- 1/4 (0.75 unities per hour)
- 1/6 (0.5 unities per hour) Journey in Rime of The Frostmaden

v 0.2.1: Add Translation Spanish (thanks @HonzoNebro)

v 0.2.0: Add Metric System Support

v 0.1.0: Initial Realise


scripts/settings.js
Hooks.once("init", () => {
    game.settings.register("travel-pace", "ForcedMarchDialog", {
        name: game.i18n.localize('TravelPace.Settings.ForcedMarchDialog'),
        hint: game.i18n.localize('TravelPace.Settings.ForcedMarchDialogHint'),
        scope: "world",
        config: true,
        default: true,
        type: Boolean
    });
    game.settings.register("travel-pace", "MetricSystem", {
        name: game.i18n.localize('TravelPace.Settings.MetricSystem'),
        hint: game.i18n.localize('TravelPace.Settings.MetricSystemHint'),
        scope: "world",
        config: true,
        default: false,
        type: Boolean
    });
    game.settings.register("travel-pace", "previewSetting", {
        name: game.i18n.localize('TravelPace.Settings.previewSetting'),
        hint: game.i18n.localize('TravelPace.Settings.previewSettingHint'),
        default: {},
        type: Object,
        scope: 'world',
        config: false
    });
});


scripts/travelpace.js
const PACENORMAL = "Normal";
class TravelPace {
    static getSceneControlButtons(buttons) {
        let tokenButton = buttons.find(b => b.name == "token")
        if (tokenButton) {
            tokenButton.tools.push({
                name: "travel-pace",
                title: game.i18n.localize('TravelPace.ButtomName'),
                icon: "fa fa-location-arrow",
                visible: game.settings.get("travel-pace", "toolbarTravelPace"),
                onClick: () => TravelPace.requestMeasure()
            });
        }
    }
    static requestMeasure() {
        //ui.notifications.warn("oi");
        if (TravelPace.requestor === undefined)
            TravelPace.requestor = new TravelPaceRequestor();
        TravelPace.requestor.render(true);
    }
}
class TravelPaceRequestor extends FormApplication {
    constructor(...args) {
        super(...args)
        game.users.apps.push(this)
    }

    static get defaultOptions() {
        const options = super.defaultOptions;
        options.title = game.i18n.localize("TravelPace.RequestorName");
        options.id = "TravelPace";
        options.template = "modules/travel-pace/templates/template.html";
        options.closeOnSubmit = true;
        options.popOut = true;
        options.width = 400;
        options.height = "auto";
        options.classes = ["TravelPace", "travel-requestor"];
        return options;
    }
    activateListeners(html) {
        super.activateListeners(html);
        this.element.find("#jspeed").change(this._onUserChange.bind(this));
        this.element.find("#jmilesonroad").change(this._onUserChange.bind(this));
        this.element.find("#jmilesoffroad").change(this._onUserChange.bind(this));

        this.element.find(".travelpace-slow-buttom").click((event) => this.setTravelPace(event, 'Slow'));
        this.element.find(".travelpace-normal-buttom").click((event) => this.setTravelPace(event, 'Normal'));
        this.element.find(".travelpace-fast-buttom").click((event) => this.setTravelPace(event, 'Fast'))
    }
    setTravelPace(event, pace) {
        event.preventDefault();
        this.element.find("#pace").val(pace);
        $("#travel-pace").submit();
    }
    JourneyTime(speed, onroad, offroad, march, outDay = false) {
        let realDistance = ((onroad * 1) + (offroad * 2));
        let realSpeed = speed / 10;
        let total = realDistance / realSpeed;
        if (march === "Slow") {
            total = realDistance / (realSpeed / 3 * 2);
        } else if (march === "Fast") {
            total = realDistance / (realSpeed / 3 * 4);
        }
        let formatMeasure = game.i18n.format("TravelPace.Dialog.Journey", { hour: Math.floor(total), min: Math.floor(((total * 60) % 60)) });
        //The Travel Pace table assumes that characters travel for 8 hours in day
        if (outDay === true) {
            let days = Math.round(total / 8);
            if (days > 1) {
                formatMeasure += game.i18n.format("TravelPace.Dialog.JourneyDays", { days: days });
            } else {
                formatMeasure += game.i18n.format("TravelPace.Dialog.JourneyDay", { days: days });
            }
            
        }
        return formatMeasure;
    }
    async getData() {
        const pwNormal = this.JourneyTime(30, 10, 10, "Normal");
        const pwSlow = this.JourneyTime(30, 10, 10, "Slow");
        const pwQuick = this.JourneyTime(30, 10, 10, "Fast");
        return { pwNormal, pwSlow, pwQuick }
    }
    _onUserChange() {
        let typeSpeed = this.element.find("#jspeed").val();
        let typeOnRoad = this.element.find("#jmilesonroad").val();
        let typeOffRoad = this.element.find("#jmilesoffroad").val();

        $("#march-normal").html(this.JourneyTime(typeSpeed, typeOnRoad, typeOffRoad, "Normal"));
        $("#march-slow").html(this.JourneyTime(typeSpeed, typeOnRoad, typeOffRoad, "Slow"));
        $("#march-fast").html(this.JourneyTime(typeSpeed, typeOnRoad, typeOffRoad, "Fast"));
    }
    async _updateObject(event, formData) {
        //TODO: Fazer as mensagens do chat.
        //ui.notifications.warn('oi2');
        let speaker = ChatMessage.getSpeaker();
        if (!speaker.actor && game.user.character) speaker = ChatMessage.getSpeaker({ actor: game.user.character });
        let templateChat = "modules/travel-pace/templates/templateChat.html";
        let marchTotal = this.JourneyTime(formData["jspeed"], formData["jmilesonroad"], formData["jmilesoffroad"], formData["pace"], true);
        let marchDisclaimer = (formData["pace"] === 'Slow') ? game.i18n.localize("TravelPace.Dialog.SlowEffects") : (formData["pace"] === 'Fast') ? game.i18n.localize("TravelPace.Dialog.FastEffects"):"";
        let dialogNarrative = game.i18n.format("TravelPace.Dialog.Narrative", { marchSpeed: formData["jspeed"], marchOnRoad: formData["jmilesonroad"], marchOffRoad: formData["jmilesoffroad"] });
        let marchstring = "TravelPace.Pace" + formData["pace"]; 
        let marchType = game.i18n.localize(marchstring);
        let dialogData = {
            marchType: marchType,
            dialogNarrative: dialogNarrative,
            marchTotal: marchTotal,
            marchDisclaimer: marchDisclaimer
        };
        let flavor = "<h3>" + game.i18n.localize("TravelPace.Dialog.Who") + "</h3>";
        let content = await renderTemplate(templateChat, dialogData);
        let messageData = {
            content: content,
            flavor: flavor,
            speaker: speaker
        };
        ChatMessage.create(messageData);
        console.log("Travel Pace submit: ", formData);
    }
}
Hooks.on('getSceneControlButtons', TravelPace.getSceneControlButtons);

scripts/travel-pace.js
const PACENORMAL = "Normal";
class TravelPace {
    static getSceneControlButtons(buttons) {
        let tokenButton = buttons.find(b => b.name == "token")
        if (tokenButton) {
            tokenButton.tools.push({
                name: "travel-pace",
                title: game.i18n.localize('TravelPace.ButtomName'),
                icon: "fa fa-location-arrow",
                visible: true,
                onClick: () => TravelPace.requestMeasure()
            });
        }
    }
    static requestMeasure() {
        //ui.notifications.warn("oi");
        if (TravelPace.requestor === undefined)
            TravelPace.requestor = new TravelPaceRequestor();
        TravelPace.requestor.render(true);
    }
}
class TravelPaceRequestor extends FormApplication {
    constructor(...args) {
        super(...args)
        game.users.apps.push(this);
    }

    static get defaultOptions() {
        const options = super.defaultOptions;
        options.title = game.i18n.localize("TravelPace.RequestorName");
        options.id = "travel-pace";
        options.template = "modules/travel-pace/templates/template.html";
        options.closeOnSubmit = true;
        options.popOut = true;
        options.width = 400;
        options.height = "auto";
        options.classes = ["travel-pace"];
        return options;
    }
    activateListeners(html) {
        super.activateListeners(html);
        this.element.find("#jspeed").change(this._onUserChange.bind(this));
        this.element.find("#jmilesonroad").change(this._onUserChange.bind(this));
        this.element.find("#jmilesoffroad").change(this._onUserChange.bind(this));
        this.element.find("#jratio").change(this._onUserChange.bind(this));

        this.element.find(".travelpace-slow-button").click((event) => this.setTravelPace(event, 'Slow'));
        this.element.find(".travelpace-normal-button").click((event) => this.setTravelPace(event, 'Normal'));
        this.element.find(".travelpace-fast-button").click((event) => this.setTravelPace(event, 'Fast'))
    }
    setTravelPace(event, pace) {
        event.preventDefault();
        this.element.find("#pace").val(pace);
        $("#travel-pace-form").submit();
    }
    JourneyTime(speed, onroad, offroad, march, ratio, outDay = false) {
        //console.log(ratio);
        if (game.settings.get("travel-pace", "MetricSystem") === true) {
            speed = Math.round(speed / 0.3);
            onroad = Math.round(onroad / 1.5);
            offroad = Math.round(offroad / 1.5);
        }
        let realDistance = ((onroad * 1) + (offroad * 2));
        let realSpeed = (speed / 10);
        let total = (realDistance / realSpeed) * ratio;        
        if (march === "Slow") {
            total = (realDistance / (realSpeed / 3 * 2)) * ratio;
        } else if (march === "Fast") {
            total = (realDistance / (realSpeed / 3 * 4)) * ratio;
        }
        let formatMeasure = game.i18n.format("TravelPace.Dialog.Journey", { hour: Math.floor(total), min: Math.floor(((total * 60) % 60)) });
        //The Travel Pace table assumes that characters travel for 8 hours in day
        if (outDay === true) {
            let days = Math.round(total / 8);
            if (days > 1) {
                formatMeasure += game.i18n.format("TravelPace.Dialog.JourneyDays", { days: days });
            } else {
                formatMeasure += game.i18n.format("TravelPace.Dialog.JourneyDay", { days: days });
            }
            
        }
        return formatMeasure;
    }

    formatMetricSystem() {
        const metricSystem = game.settings.get("travel-pace", "MetricSystem");
        let unitFeed = (metricSystem) ? game.i18n.localize("TravelPace.meters") : game.i18n.localize("TravelPace.feets");
        let unitFeedAbbr = (metricSystem) ? game.i18n.localize("TravelPace.metersAbbr") : game.i18n.localize("TravelPace.feetsAbbr");
        let unitMiles = (metricSystem) ? game.i18n.localize("TravelPace.kilometers") : game.i18n.localize("TravelPace.miles");
        let unitMilesAbbr = (metricSystem) ? game.i18n.localize("TravelPace.kilometersAbbr") : game.i18n.localize("TravelPace.milesAbbr");
        let medidas = [unitFeed, unitFeedAbbr, unitMiles, unitMilesAbbr];
        return medidas;
    }
    async getData() {
        let basemed = this.formatMetricSystem();
        let previewSetting = game.settings.get("travel-pace", "previewSetting");
        let previewSpeedVal = (game.settings.get("travel-pace", "MetricSystem")) ? 9 : 30;
        if (previewSetting.jspeed != undefined) previewSpeedVal = previewSetting.jspeed;
        let previewOnroadVal = (game.settings.get("travel-pace", "MetricSystem")) ? 5 : 3;
        if (previewSetting.jmilesonroad != undefined) previewOnroadVal = previewSetting.jmilesonroad;
        let previewOffroadVal = (game.settings.get("travel-pace", "MetricSystem")) ? 5 : 3;
        if (previewSetting.jmilesoffroad != undefined) previewOffroadVal = previewSetting.jmilesoffroad;
        let r6Select = (previewSetting.jratio == "6") ? "selected" : "";
        let r4Select = (previewSetting.jratio == "4") ? "selected" : "";
        let r2Select = (previewSetting.jratio == "2") ? "selected" : "";
        let rnSelect = (previewSetting.jratio == "1") ? "selected" : "";
        let typeRatio = (previewSetting.jratio) ? previewSetting.jratio : 1;

        const pwNormal = this.JourneyTime(previewSpeedVal, previewOnroadVal, previewOffroadVal, "Normal", typeRatio);
        const pwSlow = this.JourneyTime(previewSpeedVal, previewOnroadVal, previewOffroadVal, "Slow", typeRatio);
        const pwQuick = this.JourneyTime(previewSpeedVal, previewOnroadVal, previewOffroadVal, "Fast", typeRatio);
        const previewSpeed = game.i18n.format("TravelPace.Dialog.PreviewSpeed", { units_feetAbbr: basemed[0] });
        const previewOnroad = game.i18n.format("TravelPace.Dialog.PreviewOnroad", { units_miles: basemed[2] });
        const previewOffroad = game.i18n.format("TravelPace.Dialog.PreviewOffroad", { units_miles: basemed[2] });
        const previewRatio = game.i18n.localize("TravelPace.Dialog.PreviewRatio");
        return { pwNormal, pwSlow, pwQuick, previewSpeed, previewOnroad, previewOffroad, previewSpeedVal, previewOnroadVal, previewOffroadVal, previewRatio, rnSelect, r2Select, r4Select, r6Select }
    }
    _onUserChange() {
        let typeSpeed = this.element.find("#jspeed").val();
        let typeOnRoad = this.element.find("#jmilesonroad").val();
        let typeOffRoad = this.element.find("#jmilesoffroad").val();
        let typeRatio = this.element.find("#jratio").val();

        $("#march-normal").html(this.JourneyTime(typeSpeed, typeOnRoad, typeOffRoad, "Normal", typeRatio));
        $("#march-slow").html(this.JourneyTime(typeSpeed, typeOnRoad, typeOffRoad, "Slow", typeRatio));
        $("#march-fast").html(this.JourneyTime(typeSpeed, typeOnRoad, typeOffRoad, "Fast", typeRatio));
    }
    async _updateObject(event, formData) {
        //TODO: Fazer as mensagens do chat.
        //ui.notifications.warn('oi2');
        let speaker = ChatMessage.getSpeaker();
        // Translate metric system        
        let basemed = this.formatMetricSystem();
        // preview value
        game.settings.set("travel-pace", "previewSetting", { jspeed: formData["jspeed"], jmilesonroad: formData["jmilesonroad"], jmilesoffroad: formData["jmilesoffroad"], jratio: formData["jratio"]});

        if (!speaker.actor && game.user.character) speaker = ChatMessage.getSpeaker({ actor: game.user.character });
        let templateChat = "modules/travel-pace/templates/templateChat.html";
        let marchTotal = this.JourneyTime(formData["jspeed"], formData["jmilesonroad"], formData["jmilesoffroad"], formData["pace"], formData["jratio"], true);
        let marchDisclaimer = (formData["pace"] === 'Slow') ? game.i18n.localize("TravelPace.Dialog.SlowEffects") : (formData["pace"] === 'Fast') ? game.i18n.localize("TravelPace.Dialog.FastEffects"):"";
        let dialogNarrative = game.i18n.format("TravelPace.Dialog.Narrative", { units_miles: basemed[2], units_feetAbbr: basemed[1], marchSpeed: formData["jspeed"], marchOnRoad: formData["jmilesonroad"], marchOffRoad: formData["jmilesoffroad"] });
        let marchstring = "TravelPace.Pace" + formData["pace"]; 
        let marchType = game.i18n.localize(marchstring);
        let dialogData = {
            marchType: marchType,
            dialogNarrative: dialogNarrative,
            marchTotal: marchTotal,
            marchDisclaimer: marchDisclaimer,
            chatForced: game.settings.get("travel-pace", "ForcedMarchDialog")
        };
        let flavor = "<h3>" + game.i18n.localize("TravelPace.Dialog.Who") + " (ratio: 1/" + formData["jratio"]+")</h3>";
        let content = await renderTemplate(templateChat, dialogData);
        let messageData = {
            content: content,
            flavor: flavor,
            speaker: speaker
        };
        ChatMessage.create(messageData);
        console.log("Travel Pace submit: ", formData);
    }


}
Hooks.on('getSceneControlButtons', TravelPace.getSceneControlButtons);